### PR TITLE
ent-5767 : Kotlin 1.3/1.4 abstract classes/methods are not instrumented correctly in quasar

### PR DESCRIPTION
**Pull request by PaulHatcherR3**

### CODE REVIEW

File: quasar-core/src/main/java/co/paralleluniverse/fibers/instrument/CheckInstrumentationVisitor.java [hunk 1]
```diff
@@ -69,6 +70,7 @@
     private ClassEntry classEntry;
     private boolean hasSuspendable;
     private boolean alreadyInstrumented;
+    private ArrayList<Pair<String,String>> methodsSyntheticStatic;
```
Comment (chrisr3) [inline quasar-core/src/main/java/co/paralleluniverse/fibers/instrument/CheckInstrumentationVisitor.java:73]: Why not
```java
private final List<Pair<String, String> methodsSyntheticStatic;
```
You can then initialise it in the constructor and won't need to check whether it's been created every time. (This is for a `ClassVisitor`, so I don't see that always allocating one extra `ArrayList` per class is a hardship.)

File: quasar-core/src/main/java/co/paralleluniverse/fibers/instrument/CheckInstrumentationVisitor.java [hunk 2]
```diff
@@ -153,6 +156,17 @@ public AnnotationVisitor visitAnnotation(String adesc, boolean visible) {
...
                 public void visitEnd() {
                     super.visitEnd();
+
+                    // If we have a method not suspendable that is synthesized static method then save for later processing.
+                    // It may be missing annotations that we can fix later in visitEnd.
+                    final int ACC_STATIC_SYNTHETIC = Opcodes.ACC_STATIC | Opcodes.ACC_SYNTHETIC;
+                    if (!susp && ((access & ACC_STATIC_SYNTHETIC) == ACC_STATIC_SYNTHETIC) && name.endsWith("$default")) {
+                        if (methodsSyntheticStatic == null) {
+                            methodsSyntheticStatic = new ArrayList<>();
+                        }
```
Comment (chrisr3) [inline quasar-core/src/main/java/co/paralleluniverse/fibers/instrument/CheckInstrumentationVisitor.java:166]: Just create the list in the constructor.

File: quasar-core/src/main/java/co/paralleluniverse/fibers/instrument/CheckInstrumentationVisitor.java [hunk 3]
```diff
@@ -161,4 +175,58 @@ public void visitEnd() {
...
+            desc.append(staticMethodTypes[i].toString());
+        }
+        desc.append(")");
+        desc.append(Type.getReturnType(descStatic));
+        return desc.toString();
+    }
+
+    @Override
+    public void visitEnd() {
+        if (methodsSyntheticStatic != null) {
```
Comment (chrisr3) [inline quasar-core/src/main/java/co/paralleluniverse/fibers/instrument/CheckInstrumentationVisitor.java:204]: Just iterate over it. This loop will do nothing if the list is empty.

**Review by chrisr3 — CHANGES_REQUESTED** [review]
Small things :grinning:. My OSGi PR should merge on top with little effort too. FYI, I am trying to isolate the instrumentation code within the `co.paralleluniverse.fibers.instrument` and `co.paralleluniverse.asm` packages. I have already duplicated the `Pair` class into the `instrument` package for the sake of `waivers`.

**Review by chrisr3 — APPROVED** [review]
LGTM (Although the `import` statements look like they've been automatically updated according to two different conventions - say IntelliJ and NetBeans?)

**Comment by rick-r3** [comment]
Could you add a little bit of the detail from the JIRA to the PR (e.g. the title!) so the commit history is a little easier to understand?

Could you also prefix the PR title with ENT-5767 ?

Should just make it easier for others to follow what is going on or to identify commits in the future for you.

**Review by josephzunigadaly — APPROVED** [review]

**Review by rick-r3 — APPROVED** [review]
LGTM

**Pull request merged**